### PR TITLE
Fixing macro rendering with forms

### PR DESCRIPTION
**Pull request by Shazwazza**
Macros with forms were not working because none of current ViewData/ModelState/TempData were being carried through to the view context used to render macros.

## Main changes

* ProxyViewDataFeature now carries both ViewData and TempData so that we can use this during routing from a SurfaceController. Previously it was only carrying through ViewData
* Fixes PartialViewMacroEngine and ensure ViewData and TempData are carried across
* Fixes the ValidationSummary ext method to correctly deal with the validation ModelState prefixes

## Other

* new `UmbracoRoutedRequest` event - raised when we have a front-end routed request with an assigned content item. We already have `UmbracoRequestBegin` (which fires before routing takes place) and `UmbracoRequestEnd` which fires at the very end of the request. There was no event to know when a front-end request was attempted.
* We now use this event to ensure the `IUmbracoWebsiteSecurityAccessor` - but have added notes to this since we should kill this. The `IUmbracoWebsiteSecurity` is also just throwing exceptions since it's not complete. This is part of the members migration.
* Changes HybridAccessorBase to automatically assign a key instead of hard coding keys (these were out of date with the namespaces changes)
* Adds notes about Temporal Coupling with the IBackOfficeSecurity and UmbracoContext, we'll need to decouple these things entirely and not have one relying on transitive dependencies of the other

## Testing

* Create a new partial view macro from a snippet: use the login snippet
* Allow this macro to be used within an RTE
* Add this macro to an RTE
* Render the value in a view
* Modify your partial view .cshtml, add this somewhere to render out the TempData value when its successful. Currently the UmbLoginController just sets this to `true` (not actually sure why)
   ```cs
   if (TempData.TryGetValue("LoginSuccess", out var msg))
   {
       <div>@msg</div>
   }
   ```
* Add breakpoint to UmbLoginController since `_websiteSecurityAccessor.WebsiteSecurity.LoginAsync` is just going to throw... so you can manually change how the result works, either move the execution to the `AddModelError` to add an error and return once that is tested, test moving the execution to `TempData["LoginSuccess"] = true;` for a successful test. 
* You should see that the validation summary and TempData behave correctly

Bonus points:

* Create a custom ViewComponent that looks up the IPublishedContent item above which renders the RTE
* Then render this ViewComponent in a view of a different IPublishedContent
* Now we have multiple levels of ViewComponents and controllers
* The ViewData/ModelState/TempData should still work as expected

### CODE REVIEW

**Review by bergmania — COMMENTED** [review]
When I test it seems to work 🎉 

Only a few comments to the code, but what to do with all the TODOs. 

I think the reason the namespace is wrong is due to a merge error, and the `*.csproj` contains multiple RootNamespace elements

**Comment by bergmania** [comment]
Btw, I can do the changes if you agree, so you don't have to context switch

File: src/Umbraco.Core/HybridAccessorBase.cs [hunk 1]
```diff
@@ -18,12 +18,15 @@ public abstract class HybridAccessorBase<T>
     {
         private readonly IRequestCache _requestCache;
 
+        // TODO: Do they need to be static?? These are singleton instances IMO they shouldn't be static
```
[the remark this answers is not in the document]
Comment (Shazwazza) [reply src/Umbraco.Core/HybridAccessorBase.cs:21]: Yep cool lets change them

File: src/Umbraco.Core/HybridAccessorBase.cs [hunk 2]
```diff
@@ -43,34 +46,42 @@ public abstract class HybridAccessorBase<T>
...
             {
-                if (o == null) return;
-                var value = o as T;
-                if (value == null) throw new ArgumentException($"Expected type {typeof(T).FullName}, got {o.GetType().FullName}", nameof(o));
+                if (o == null)
+                {
+                    return;
+                }
+
+                if (!(o is T value))
```
[the remark this answers is not in the document]
Comment (Shazwazza) [reply src/Umbraco.Core/HybridAccessorBase.cs:80]: Sounds awesome, been wondering when we'd start supporting it

File: src/Umbraco.Core/Security/IUmbracoWebsiteSecurity.cs [hunk 3]
```diff
@@ -1,9 +1,10 @@
-﻿using System.Collections.Generic;
+using System.Collections.Generic;
 using System.Threading.Tasks;
 using Umbraco.Cms.Core.Models.Security;
 
 namespace Umbraco.Cms.Core.Security
 {
+    // TODO: I think we can kill this whole thing, the logic can just be in the controllers
```
[the remark this answers is not in the document]
Comment (Shazwazza) [reply src/Umbraco.Core/Security/IUmbracoWebsiteSecurity.cs:7]: Yes, I would 100% like to kill IBackOfficeSecurity too. If we want one abstraction to do what both MemberManager/MemberSignInManager (and similarly for users) does then this diverts from normal ASP.NET Identity behavior. But if we really still want to do that, then we can potentially have the sign in methods available on our manager abstraction - but I foresee problems with circular refs. Both IBackOfficeSecurity  and IUmbracoWebsiteSecurity are from legacy code where we had a huge mix of auth types and strange methods.

File: src/Umbraco.Web.Common/Macros/MacroRenderer.cs [hunk 4]
```diff
@@ -338,38 +341,28 @@ private Attempt<MacroContent> ExecuteProfileMacroWithErrorWrapper(MacroModel mac
...
             // ensure that we are running against a published node (ie available in XML)
             // that may not be the case if the macro is embedded in a RTE of an unpublished document
 
             if (content == null)
+            {
                 return Attempt.Fail(new MacroContent { Text = "[macro failed (no content)]" });
+            }
 
-            var textService = _textService;
+            ILocalizedTextService textService = _textService;
```
[the remark this answers is not in the document]
Comment (Shazwazza) [reply src/Umbraco.Web.Common/Macros/MacroRenderer.cs:356]: Agree, this must be left over from some refactoring

**Comment by Shazwazza** [comment]
> Btw, I can do the changes if you agree, so you don't have to context switch

That would be super if you have time! I've left some minor comments.

File: src/Umbraco.Web.UI.NetCore/Umbraco.Web.UI.NetCore.csproj [hunk 5]
```diff
@@ -21,6 +21,7 @@
 
   <ItemGroup>
     <Folder Include="App_Plugins" />
+    <Folder Include="Views" />
```
Comment (Shazwazza) [inline src/Umbraco.Web.UI.NetCore/Umbraco.Web.UI.NetCore.csproj]: Don't think we want this one do we? won't cause issues for now though. Everything looks good, i'll merge

**Pull request merged**